Content.Server/Chemistry/EntitySystems/ReagentDispenserSystem.cs

```diff
@@ -1,32 +1,152 @@
+        [Dependency] private readonly SolutionContainerSystem _solutionContainerSystem = default!;
+        [Dependency] private readonly ItemSlotsSystem _itemSlotsSystem = default!;
+        [Dependency] private readonly UserInterfaceSystem _userInterfaceSystem = default!;
+        [Dependency] private readonly IPrototypeManager _prototypeManager = default!;
+        [Dependency] private readonly IAdminLogManager _adminLogger = default!;
+        [Dependency] private readonly EntityManager _entities = default!;
         public override void Initialize()
         {
             base.Initialize();
-            SubscribeLocalEvent<ReagentDispenserComponent, ComponentStartup>((_, comp, _) => comp.UpdateUserInterface());
-            SubscribeLocalEvent<ReagentDispenserComponent, SolutionChangedEvent>((_, comp, _) => comp.UpdateUserInterface());
-            SubscribeLocalEvent<ReagentDispenserComponent, EntInsertedIntoContainerMessage>((_, comp, _) => comp.UpdateUserInterface());
-            SubscribeLocalEvent<ReagentDispenserComponent, EntRemovedFromContainerMessage>((_, comp, _) => comp.UpdateUserInterface());
+
+            SubscribeLocalEvent<ReagentDispenserComponent, ComponentStartup>((_, comp, _) => UpdateUiState(comp));
+            SubscribeLocalEvent<ReagentDispenserComponent, SolutionChangedEvent>((_, comp, _) => UpdateUiState(comp));
+            SubscribeLocalEvent<ReagentDispenserComponent, EntInsertedIntoContainerMessage>((_, comp, _) => UpdateUiState(comp));
+            SubscribeLocalEvent<ReagentDispenserComponent, EntRemovedFromContainerMessage>((_, comp, _) => UpdateUiState(comp));
+            SubscribeLocalEvent<ReagentDispenserComponent, BoundUIOpenedEvent>((_, comp, _) => UpdateUiState(comp));
             SubscribeLocalEvent<ReagentDispenserComponent, GotEmaggedEvent>(OnEmagged);
+
+            SubscribeLocalEvent<ReagentDispenserComponent, ReagentDispenserSetDispenseAmountMessage>(OnSetDispenseAmountMessage);
+            SubscribeLocalEvent<ReagentDispenserComponent, ReagentDispenserDispenseReagentMessage>(OnDispenseReagentMessage);
+            SubscribeLocalEvent<ReagentDispenserComponent, ReagentDispenserClearContainerSolutionMessage>(OnClearContainerSolutionMessage);
+        }
+
+        private void UpdateUiState(ReagentDispenserComponent reagentDispenser)
+        {
+            if (!HasComp<ItemSlotsComponent>(reagentDispenser.Owner))
+                return;
+
+            var outputContainer = _itemSlotsSystem.GetItem(reagentDispenser.Owner, SharedReagentDispenser.OutputSlotName);
+            var outputContainerInfo = BuildOutputContainerInfo(outputContainer);
+
+            var dispenserName = Name(reagentDispenser.Owner);
+            var inventory = GetInventory(reagentDispenser);
+
+            var state = new ReagentDispenserBoundUserInterfaceState(Name(reagentDispenser.Owner), outputContainerInfo, inventory, reagentDispenser.DispenseAmount);
```

Comment:
`dispenserName` is unused. dont think we even really need to send the name? pretty sure the client can just set the title itself

Reply:
Fixed the duplicate `Name()` call. I'm not sure how to get the entity name from inside the UI code, and the default is "Reagent dispenser" for the chem dispenser, soda dispenser, etc which is not ideal. So it still sends the entity name from the server.

Reply:
I figured out how to get entity names from inside the `BoundUserInterface` and set the title that way. I applied the same fix to chemmaster also.